src/Stream/MessageWriter.h

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This code has been produced by the European Spallation Source
+// and its partner institutes under the BSD 2 Clause License.
+//
+// See LICENSE.md at the top level for license information.
+//
+// Screaming Udder!                              https://esss.se
+
+/// \brief Do the actual file writing.
+///
+
+#pragma once
+
+#include "Message.h"
+#include "Metrics/Metric.h"
+#include "Metrics/Registrar.h"
+#include "ThreadedExecutor.h"
+#include "logger.h"
+#include <map>
+#include <thread>
+
+namespace WriterModule {
+class Base;
+}
+
+namespace Stream {
+
+class MessageWriter {
+public:
+  explicit MessageWriter(Metrics::Registrar const &MetricReg);
+
+  virtual void addMessage(Message const &Msg);
+
+  using ModuleHash = size_t;
+
+protected:
+  void writeMsgImpl(WriterModule::Base *ModulePtr,
+                    FileWriter::FlatbufferMessage const &Msg);
+
+  SharedLogger Log{getLogger()};
+  Metrics::Metric WritesDone{"writes_done",
+                             "Number of completed writes to HDF file."};
+  Metrics::Metric WriteErrors{"write_errors",
+                              "Number of failed HDF file writes.",
+                              Metrics::Severity::ERROR};
+  std::map<ModuleHash, std::unique_ptr<Metrics::Metric>> ModuleErrorCounters;
```

Comment:
I think these three items can either be public or have a public accessor so the tests can get to these counters without the 'using' workaround.
It is fine line, but it is often reasonable to add things like these to the public API to make testing easier.

Reply:
These member variables should really not be public as it would then be way too easy to mess upp the metrics and using public accessors adds otherwise unnecessary code that would only be used by the unit tests (for now at least). Seeing as you say it is a fine line, do you want to have the public accessors?

Reply:
If it is a choice between public accessors or messing about with inheritance and `using` I will (reluctantly) take the accessors.
The other (better) option would be to get the results through the mock registrar, if that was possible?

Reply:
Getting them through the metrics registrar is definitely possible but that would be even more code that would then only be used for these unit tests. Probably on the order 50 to 100 lines for implementing the possibility of doing this. In short, I REALLY do not think it is worth that effort right now.